java/vector/src/main/java/org/apache/arrow/vector/BitVector.java

```diff
@@ -276,36 +276,55 @@ public void transferTo(BitVector target) {
 
   public void splitAndTransferTo(int startIndex, int length, BitVector target) {
     assert startIndex + length <= valueCount;
-    int firstByte = getByteIndex(startIndex);
-    int byteSize = getSizeFromCount(length);
+    int firstByteSource = getByteIndex(startIndex);
+    int lastByteSource = getByteIndex(valueCount-1);
+    int byteSizeTarget = getSizeFromCount(length);
     int offset = startIndex % 8;
+
     if (length > 0) {
       if (offset == 0) {
         target.clear();
         // slice
         if (target.data != null) {
           target.data.release();
         }
-        target.data = data.slice(firstByte, byteSize);
+        target.data = data.slice(firstByteSource, byteSizeTarget);
         target.data.retain(1);
-      } else {
+      }
+      else {
         // Copy data
         // When the first bit starts from the middle of a byte (offset != 0), copy data from src BitVector.
         // Each byte in the target is composed by a part in i-th byte, another part in (i+1)-th byte.
-        // The last byte copied to target is a bit tricky :
-        //   1) if length requires partly byte (length % 8 !=0), copy the remaining bits only.
-        //   2) otherwise, copy the last byte in the same way as to the prior bytes.
+
         target.clear();
-        target.allocateNew(length);
+        target.allocateNew(byteSizeTarget * 8);
+        int destPos = 0;
+        int sourcePos = startIndex;
         // TODO maybe do this one word at a time, rather than byte?
-        for (int i = 0; i < byteSize - 1; i++) {
-          target.data.setByte(i, (((this.data.getByte(firstByte + i) & 0xFF) >>> offset) + (this.data.getByte(firstByte + i + 1) << (8 - offset))));
+
+        int counter = firstByteSource;
+
+        for (int i = 0; i < byteSizeTarget - 1; i++) {
+          target.data.setByte(i,
+                  (((this.data.getByte(firstByteSource + i) & 0xFF) >>> offset) + (this.data.getByte(firstByteSource + i + 1) << (8 - offset))));
+          counter++;
         }
-        if (length % 8 != 0) {
-          target.data.setByte(byteSize - 1, ((this.data.getByte(firstByte + byteSize - 1) & 0xFF) >>> offset));
-        } else {
-          target.data.setByte(byteSize - 1,
-              (((this.data.getByte(firstByte + byteSize - 1) & 0xFF) >>> offset) + (this.data.getByte(firstByte + byteSize) << (8 - offset))));
+
+        /* Copying the last piece is done in the following manner:
+         * if the source vector has 1 or more bytes remaining, we copy the last piece as a byte formed by shifting data
+         * from the current byte and the next byte.
+         *
+         * if the source vector has no more bytes remaining (we are at the last byte), we copy the last piece as a byte
+         * by shifting data from the current byte.
+         */
+        if((counter < lastByteSource)) {
```

Comment:
Just `if (counter < lastByteSource)`? (remove extra parenthesis)

Reply:
Also, I think `firstByteSource + byteSizeTarget - 1 < lastByteSource` is better than `count < lastByteSource` because we don't need to worry about making sure the counter is correct.

I also prefer `byteSizeToCopy` to `byteSizeTarget` a bit better but that's very minor. So up to you.